Documentation/bpf.rst

```diff
@@ -0,0 +1,1855 @@
+BPF samples in the kernel source tree (``samples/bpf/``) that parse the generated
+object files and load the code directly through the system call interface.
+
+This and previous sections mainly focus on the iproute2 suite's eBPF front end for
+loading networking programs of XDP, tc or lwt type, since Cilium's programs are
+implemented against this eBPF loader. In future, Cilium will ship with a native
+eBPF loader, but programs will still be compatible to be loaded through iproute2
+suite in order to facilitate development and debugging.
+
+All eBPF program types supported by iproute2 share the same eBPF loader logic
+due to having a common loader back end implemented as a library (``lib/bpf.c``
+in iproute2 source tree).
+
+The previous section on LLVM also covered some iproute2 parts related to writing
+eBPF C programs, and later sections in this document are related to tc and XDP
+specific aspects when writing programs. Therefore, this section will rather focus
+on usage examples for loading object files with iproute2 as well as some of the
+generic mechanics of the loader. It does not try to provide a complete coverage
+of all details, but enough for getting started.
+
+**1. Loading of XDP eBPF object files.**
+
+  Given an eBPF object file ``prog.o`` has been compiled for XDP, it can be loaded
+  through ``ip`` to a XDP-supported netdevice called ``em1`` with the following
+  command:
+
+  ::
+
+    # ip link set dev em1 xdp obj prog.o
+
+  The above command assumes that the program code resides in the default section
+  which is called ``prog`` in XDP case. Should this not be the case, and the
+  section named differently, for example, ``foobar``, then the program needs to
+  be loaded as:
+
+  ::
+
+    # ip link set dev em1 xdp obj prog.o sec foobar
+
+  By default, ``ip`` will throw an error in case a XDP program is already attached
+  to the networking interface, thus that it will not be overridden by accident. In
+  order to replace the currently running XDP program with a new one, the ``-force``
+  option must be used:
+
+  ::
+
+    # ip -force link set dev em1 xdp obj prog.o
+
+  Most XDP-enabled drivers today support an atomic replacement of the existing
+  program with a new one without traffic interruption. There is always only a
+  single program attached to an XDP-enabled driver due to performance reasons,
+  hence a chain of programs is not supported. However, as described in the
+  previous section, partitioning of programs can be performed through tail
+  calls to achieve a similar use-case when necessary.
+
+  The ``ip link`` command will display an ``xdp`` flag if the interface has an XDP
+  program attached. ``ip link | grep xdp`` can thus be used to find all interfaces
+  that have XDP running. Further introspection facilities will be provided through
+  the detailed view through ``ip -d link`` once the kernel API gains support for
+  dumping additional attributes.
```

Comment:
s/view through/view/ ? The extra `through` adds nothing extra IMO, but either way is fine.